### PR TITLE
Python: Add 6962-bis data structures.

### PR DESCRIPTION
**Pull request by eranmes**
This is based on draft 20.

### CODE REVIEW

File: python/ct/proto/client_v2.proto [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+enum SctExtensionType {
+  option (tls_enum_opts).max_value = 65535;
+
+  SCT_EXTENSION_RESERVED=65535;
+}
+
+message SctExtensionV2 {
+  optional SctExtensionType type = 1;
+  optional bytes sct_extension_data = 2 [(tls_opts).min_length = 1,
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: This should be `min_length=0`; https://tools.ietf.org/html/draft-ietf-trans-rfc6962-bis-20#section-5.6

File: python/ct/proto/client_v2.proto [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+message PrecertChainEntryV2 {
+  // pre_certificate, opaque (CMSPrecert) <1..2^24-1>
+  optional bytes pre_certificate = 1 [
+    (tls_opts).min_length = 1,
+    (tls_opts).max_length = 0xffffff ];
+
+  // precertificate_chain, repeated opaque (ASN.1Cert) <1..2^24-1>
+  // A chain from the precert to a trusted root, excluding the precert itself.
+  // Must have at least one certificate.
+  repeated bytes precertificate_chain = 2 [
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: `pre_certificate_chain` would be more consistent with the field above and with 6962-bis

File: python/ct/proto/client_v2.proto [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+enum SctExtensionType {
+  option (tls_enum_opts).max_value = 65535;
+
+  SCT_EXTENSION_RESERVED=65535;
+}
+
+message SctExtensionV2 {
+  optional SctExtensionType type = 1;
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: `sct_extension_type` would be more consistent with the field below and with the draft.

File: python/ct/proto/client_v2.proto [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  optional bytes sct_extension_data = 2 [(tls_opts).min_length = 1,
+                                      (tls_opts).max_length = 65535];
+}
+
+message SignedCertificateTimestampDataV2 {
+  optional LogIDV2 id = 1;
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Maybe `log_id` to keep a consistent mapping from -bis names to field names?

File: python/ct/proto/client_v2.proto [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  // UTC time in milliseconds, since January 1, 1970, 00:00.
+  optional uint64 timestamp = 2;
+  repeated SctExtensionV2 extensions = 3 [
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Ditto `sct_extensions`

File: python/ct/proto/client_v2.proto [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    (tls_opts).min_total_length = 0,
+    (tls_opts).max_total_length = 0xffff];
+  optional DigitallySigned signature = 4;
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Be nice to have a comment indicating what this is a signature over.

File: python/ct/proto/client_v2.proto [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+enum SthExtensionType {
+  option (tls_enum_opts).max_value = 65535;
+
+  STH_EXTENSION_RESERVED = 65535;
+}
+
+message SthExtension {
+  optional SthExtensionType type = 1;
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: `sth_extension_type` for consistency.

File: python/ct/proto/client_v2.proto [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+message TimestampedCertificateEntryDataV2 {
+  optional uint64 timestamp = 1;
+  optional bytes issuer_key_hash = 2 [
+    (tls_opts).min_length = 32,
+    (tls_opts).max_length = 0xff];
+  optional bytes tbs_certificate = 3 [
+    (tls_opts).min_length = 1,
+    (tls_opts).max_length = 0xffffff ];
+  repeated SctExtensionV2 sct_extensions = 4 [
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Maybe add a comment about required ordering of extensions

File: python/ct/proto/client_v2.proto [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                                      (tls_opts).max_length = 65535];
+}
+
+message TreeHeadDataV2 {
+  optional uint64 timestamp = 1;
+  optional uint64 tree_size = 2;
+  optional bytes root_hash = 3 [
+    (tls_opts).min_length = 32,
+    (tls_opts).max_length = 0xff];
+  repeated SthExtension sth_extensions = 4 [
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Comment about required ordering?

File: python/ct/proto/client_v2.proto [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    (tls_opts).min_total_length = 0,
+    (tls_opts).max_total_length = 0xffff];
+}
+
+message SignedTreeHeadDataV2 {
+  optional LogIDV2 log_id = 1;
+  optional TreeHeadDataV2 tree_head = 2;
+  optional DigitallySigned signature = 3;
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Comment about signature over what.

File: python/ct/proto/client_v2.proto [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+message ConsistencyProofDataV2 {
+  optional LogIDV2 log_id = 1;
+  optional uint64 tree_size_1 = 2;
+  optional uint64 tree_size_2 = 3;
+  repeated bytes consistency_path = 4 [
+    (tls_opts).min_length = 32,
+    (tls_opts).max_length = 0xff,
+    (tls_opts).min_total_length = 32,
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: The draft has `min_total_length` of 1, which might be an buglet in the draft.

https://tools.ietf.org/html/draft-ietf-trans-rfc6962-bis-20#section-5.9

File: python/ct/proto/client_v2.proto [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+message InclusionProofDataV2 {
+  optional LogIDV2 log_id = 1;
+  optional uint64 tree_size = 2;
+  optional uint64 leaf_index = 3;
+  repeated bytes inclusion_path = 4 [
+    (tls_opts).min_length = 32,
+    (tls_opts).max_length = 0xff,
+    (tls_opts).min_total_length = 32,
```
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Again, draft has `min_total_length` of 1.

https://tools.ietf.org/html/draft-ietf-trans-rfc6962-bis-20#section-5.10

File: python/ct/proto/client_v2.proto (hunk 2 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Not sure I understand why: In 6962-bis, the PrecertChainEntryV2 contains precertificate_chain and pre_certificate fields.
(are you suggesting we change that in 6962-bis to pre_certificate_chain?)

File: python/ct/proto/client_v2.proto (hunk 3 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Done.

File: python/ct/proto/client_v2.proto (hunk 4 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Done.

File: python/ct/proto/client_v2.proto (hunk 5 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Done.

File: python/ct/proto/client_v2.proto (hunk 6 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Done

File: python/ct/proto/client_v2.proto (hunk 7 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Done.

File: python/ct/proto/client_v2.proto (hunk 9 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Done

File: python/ct/proto/client_v2.proto (hunk 10 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Done.

File: python/ct/proto/client_v2.proto (hunk 11 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: This is intentional - no point in providing an empty consistency path (i.e., the only case where the consistency path would be empty is for a consistency proof between identical trees or between the empty tree and a non-empty tree, for both cases it makes no sense to request a consistency proof).

File: python/ct/proto/client_v2.proto (hunk 12 above)
Comment (eranmes) [inline python/ct/proto/client_v2.proto]: Again, I believe this is intentional. @robstrad could comment.

**Comment by eranmes** [comment]
David, PTAL.

File: python/ct/proto/client_v2.proto (hunk 11 above)
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: Sure, but it just seems slightly odd to set a (non-zero) minimum length that's not actually the smallest size that can occur in this field -- if it's non-zero it has to be at least 32 bytes, as in the proto attributes.

File: python/ct/proto/client_v2.proto (hunk 2 above)
Comment (daviddrysdale) [inline python/ct/proto/client_v2.proto]: You're right, not sure what I was thinking.

**Pull request merged**